### PR TITLE
Move Property implementation out of headers.

### LINKED ISSUE #17859: Move Property template implementations from headers into sources
This could be another small step in improving compile times.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is a small step that helps to reduce the number of includes. The templates `PropertyWithValue` and `WorkspaceProperty` are now not defined in their header files anymore. Instead we are using explicit instantiations.

These changes reduce the average number of distinct recursively included files for `cpp` files in `Framework/` from 110 to 107.

**To test:**

In principle passing builds and tests should indicate whether this is ok or not, but I'd prefer if someone could do a code review.

Fixes #17859.

Internal change, no release notes.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
@OwenArnold I resolved some merge conflicts, if builds pass I think this is ready.

File: Framework/Algorithms/src/NormaliseByCurrent.cpp [hunk 1]
```diff
@@ -42,15 +44,14 @@ double NormaliseByCurrent::extractCharge(
   int nPeriods = 0;
   try {
     Property *nPeriodsProperty = run.getLogData("nperiods");
-    Kernel::toValue<int>(nPeriodsProperty->value(), nPeriods);
+    nPeriods = boost::lexical_cast<int>(nPeriodsProperty->value());
```
Comment (OwenArnold) [inline Framework/Algorithms/src/NormaliseByCurrent.cpp]: with c++11 you can also now do this with [tostring](http://en.cppreference.com/w/cpp/string/basic_string/to_string) but I don'r recommend changing anything.

**Comment by OwenArnold** [comment]
These changes make sense to me. In the old days, when the framework was perceived to allow for user-provided workspaces and property types, the header implementations made sense. However, this is not the reality for the c++ side of the framework, so this is a better tradeoff.

**Comment by OwenArnold** [comment]
:shipit:

**Comment by rosswhitfield** [comment]
This has broken building with clang on Linux, you can't assume `__clang__` means OSX. I have a fix here #18213

**Pull request merged**